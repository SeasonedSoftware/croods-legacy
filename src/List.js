import { Component } from 'react'
import { connect } from 'react-redux'

import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import renderIfPresent from './renderIfPresent'

class List extends Component {
  constructor(props) {
    super(props)

    const { list, parentId, path, listPath, actions } = props

    if (!list || parentId || path !== listPath) {
      actions.fetchList(path)
    }
  }

  componentDidUpdate(prevProps) {
    const { parentId, path, actions } = this.props
    const { parentId: oldId, path: oldPath } = prevProps

    const parentChanged = parentId && parentId.toString() !== oldId.toString()
    const pathChanged = path && path.toString() !== oldPath.toString()

    if (parentChanged || pathChanged) {
      actions.fetchList(path)
    }
  }

  renderLoading() {
    const { renderLoading, list, fetchingList } = this.props

    if (renderLoading && (!list || fetchingList)) {
      return renderLoading()
    }
  }

  render() {
    const { render, list, listError, renderError } = this.props

    return (
      renderIfPresent(renderError, listError) ||
      this.renderLoading() ||
      render(list, this.props)
    )
  }
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(List),
)
