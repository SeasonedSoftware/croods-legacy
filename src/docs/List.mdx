---
name: List
route: /list
order: 10
---

import { Playground } from 'docz'

# List

The List component provides a way to fetch a list from an api, using the response as needed, fetching or set a resource and show or use the response data.

## API

| Property      | Type          | Required | Default         | Description                                                                                             |
| ------------- | ------------- | -------- | --------------- | ------------------------------------------------------------------------------------------------------- |
| name          | String        | true     | -               | It defines which reducer will be used and, if no path have been provider, will set path to baseUrl/name |
| render        | Func          | true     | -               | A function returning a React Node. Ex: ({ list }, props) -> < JSX />                                    |
| renderError   | Func          | false    | `from provider` | A custom function to be called when loading a component.                                                |
| renderLoading | Func          | false    | `from provider` | A function to be called when a error occurs.                                                            |
| parentId      | String/Number | false    | -               | Parent element's Id, for more specificity in changes.                                                   |
| path          | String        | false    | `/${name}`      | Ex: /foo/bar-\_12?&=                                                                                    |

## How to use

```javascript
import { List } from 'croods'

export default (
  <List
    name="pets"
    // if path is different than /pets, you can set a custom path like so:
    // path='/customPath'
    render={list => <Component list={list} />}
  />
)
```

### render

The render function is used to receive the api response in the form of an array, **list** , and
pass it to render in any component you like.

### renderError

The renderError function is fired when the server response has a 4xx or 5xx status.
It receives the `error` parameter containing the error message.

### renderLoading

The renderLoading function is fired while there is no response from server.

## Example

You can see a working example using this component in [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/List.js).
