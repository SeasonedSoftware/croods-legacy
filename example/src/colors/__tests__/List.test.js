import React from 'react'
import renderer from 'react-test-renderer'

import List from '../List'

jest.mock('croods', () => ({
  List: props => (
    <div {...props}>
      List -{' '}
      {props.render([{ id: 1, name: 'foo' }, { id: 2, name: 'bar' }, { id: 3, name: 'foobar' }])}
    </div>
  ),
}))

jest.mock('react-router-dom/Link', () => props => <div {...props}>Link - {props.children}</div>)

jest.mock('../ListItem', () => props => <div {...props}>ListItem - {props.children}</div>)

it('renders correctly', () => {
  const tree = renderer.create(<List />).toJSON()
  expect(tree).toMatchSnapshot()
})
