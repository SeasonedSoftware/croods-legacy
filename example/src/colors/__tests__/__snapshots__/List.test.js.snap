// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  name="colors"
  render={[Function]}
>
  List -
   
  <h1>
    Colors
  </h1>
  <div
    item={
      Object {
        "id": 1,
        "name": "foo",
      }
    }
  >
    ListItem - 
  </div>
  <div
    item={
      Object {
        "id": 2,
        "name": "bar",
      }
    }
  >
    ListItem - 
  </div>
  <div
    item={
      Object {
        "id": 3,
        "name": "foobar",
      }
    }
  >
    ListItem - 
  </div>
  <div
    style={
      Object {
        "display": "block",
        "marginTop": 20,
      }
    }
    to="/new"
  >
    Link - 
    New
  </div>
</div>
`;
